pkg/reconciler/sample/samplesource.go

```diff
@@ -47,37 +53,45 @@ var _ reconcilersamplesource.Interface = (*Reconciler)(nil)
 
 // ReconcileKind implements Interface.ReconcileKind.
 func (r *Reconciler) ReconcileKind(ctx context.Context, src *v1alpha1.SampleSource) pkgreconciler.Event {
-	ra, event := r.dr.ReconcileDeployment(ctx, src, resources.MakeReceiveAdapter(&resources.ReceiveAdapterArgs{
-		EventSource:    src.Namespace + "/" + src.Name,
-		Image:          r.ReceiveAdapterImage,
-		Source:         src,
-		Labels:         resources.Labels(src.Name),
-		AdditionalEnvs: r.configAccessor.ToEnvVars(), // Grab config envs for tracing/logging/metrics
-	}))
+
+	ctx = sourcesv1.WithURIResolver(ctx, r.sinkResolver)
+
+	ra, sb, event := r.dr.ReconcileDeployment(ctx, src, makeSinkBinding(src),
+		resources.MakeReceiveAdapter(&resources.ReceiveAdapterArgs{
+			EventSource:    src.Namespace + "/" + src.Name,
+			Image:          r.ReceiveAdapterImage,
+			Source:         src,
+			Labels:         resources.Labels(src.Name),
+			AdditionalEnvs: r.configAccessor.ToEnvVars(), // Grab config envs for tracing/logging/metrics
+		}),
+	)
 	if ra != nil {
 		src.Status.PropagateDeploymentAvailability(ra)
 	}
+	if sb != nil {
+		if c := sb.Status.GetCondition(sourcesv1.SinkBindingConditionSinkProvided); c.IsTrue() {
+			src.Status.MarkSink(sb.Status.SinkURI)
+		} else if c.IsFalse() {
+			src.Status.MarkNoSink(c.GetReason(), "%s", c.GetMessage())
+		}
+	}
 	if event != nil {
 		logging.FromContext(ctx).Infof("returning because event from ReconcileDeployment")
 		return event
 	}
 
-	if ra != nil {
-		logging.FromContext(ctx).Info("going to ReconcileSinkBinding")
-		sb, event := r.sbr.ReconcileSinkBinding(ctx, src, src.Spec.SourceSpec, tracker.Reference{
-			APIVersion: appsv1.SchemeGroupVersion.String(),
-			Kind:       "Deployment",
-			Namespace:  ra.Namespace,
-			Name:       ra.Name,
-		})
-		logging.FromContext(ctx).Infof("ReconcileSinkBinding returned %#v", sb)
-		if sb != nil {
-			src.Status.MarkSink(sb.Status.SinkURI)
-		}
-		if event != nil {
-			return event
-		}
-	}
-
 	return nil
 }
+
+func makeSinkBinding(src *v1alpha1.SampleSource) *sourcesv1.SinkBinding {
+	return &sourcesv1.SinkBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			// this is necessary to track the change of sink reference.
+			Name:      src.GetName(),
+			Namespace: src.GetNamespace(),
```

Comment:
I have an itchy feeling for this part.
Any idea to propagate namespasedkey to `URIFromDestinationV1` in `Do()`?
cc @lberk 
/hold

Reply:
what is hacky about it? seems like the way

Reply:
Well, I thought this might be tricky if seeing only the surface, but if folks can see what is going on inside, it is okay.
/unhold